### PR TITLE
Course model part2

### PR DESCRIPTION
**Pull request by Bjvanminnen**
New attempt at https://github.com/code-dot-org/code-dot-org/pull/14805/files based on feedback.

This comes in three parts (this is Part 2). Part 1 needs to be in a separate DTP, since changes to the model in future parts cause the migration not work. Parts 2 and 3 could theoretically be combined.

(1) Creates the new table, and populates it with values from Plc::Course table. The model is not used outside of the migration at all.
https://github.com/code-dot-org/code-dot-org/pull/14850
**(2) Add our association between models, delegate name from Plc::Course to Course, and update code. At this point the name column still exists on Plc::Course table, but should be unused.**
(3) Our only destructive migration, this just removes the name column.
https://github.com/code-dot-org/code-dot-org/pull/14853

I've validated that at each step along the way dashboard tests pass.

Future work (after parts 1 - 3):
Update routes.rb to have /course/:courseId go to a courses controller. This controller will then route plc courses to the appropriate plc controller
Another migration that adds a CourseScript model (representing the scripts that are in a non-plc course)

### CODE REVIEW

File: dashboard/app/controllers/plc/user_course_enrollments_controller.rb [hunk 1]
```diff
@@ -5,7 +5,7 @@ class Plc::UserCourseEnrollmentsController < ApplicationController
   def index
     if user_course_enrollment_params[:course]
       course_name = user_course_enrollment_params[:course].tr('-', '_').titleize
-      @user_course_enrollments = [@user_course_enrollments.find_by(user: current_user, plc_course: Plc::Course.find_by(name: course_name))]
+      @user_course_enrollments = [@user_course_enrollments.find_by(user: current_user, plc_course: ::Course.find_by_name(course_name).try(:plc_course))]
```
Comment (Bjvanminnen) [inline dashboard/app/controllers/plc/user_course_enrollments_controller.rb:8]: It's a little disconcerting to me that Using the syntax `Course::find_by_name...` worked in tests but not when I actually hit this code running my local server (needed to be `::Course.find_by_name`)

**Comment by Bjvanminnen** [comment]
Circle failure is in `plc/courseUnitNavigation.feature`, which is a little worrying. However, this passes locally for me, and when I investigated, it looked like the problem was actually in trying to create a user (not anything PLC specific). I'm rebuilding without cache now, and will be sure not to merge until I understand that failure.

File: dashboard/app/helpers/plc/courses_helper.rb [hunk 2]
```diff
@@ -1,5 +1,5 @@
 module Plc::CoursesHelper
   def options_for_user_enrollment_courses
-    Plc::Course.pluck(:name, :id).sort
+    Plc::Course.all.map {|c| [c.name, c.id]}.sort
```
Comment (aoby) [inline dashboard/app/helpers/plc/courses_helper.rb:3]: These are equivalent:
```
irb(main):010:0> Plc::Course.all.map{|c| [c.name, c.id]}.sort == Plc::Course.pluck(:name, :id).sort
=> true
```

Why the change? I find the former (slightly) more readable

File: dashboard/app/models/pd/workshop.rb [hunk 3]
```diff
@@ -481,7 +481,7 @@ def workshop_dashboard_url
   end
 
   def associated_online_course
-    Plc::Course.find_by(name: WORKSHOP_COURSE_ONLINE_LEARNING_MAPPING[course]) if WORKSHOP_COURSE_ONLINE_LEARNING_MAPPING[course]
+    ::Course.find_by(name: WORKSHOP_COURSE_ONLINE_LEARNING_MAPPING[course]).try(:plc_course) if WORKSHOP_COURSE_ONLINE_LEARNING_MAPPING[course]
```
Comment (aoby) [inline dashboard/app/models/pd/workshop.rb:484]: Don't need `try` and the trailing conditional. Pick one :) (I'd favor try)

File: dashboard/test/factories.rb [hunk 4]
```diff
@@ -5,7 +5,6 @@
   factory :course do
     name "MyCourseName"
     properties nil
-    plc_course_id nil
```
Comment (aoby) [inline dashboard/test/factories.rb:8]: unrelated to this PR, but we don't need to supply nil values to FactoryGirl unless we want to explicitly override a default in the model with nil. Like 7 can also be removed.

File: dashboard/app/helpers/plc/courses_helper.rb (hunk 2 above)
Comment (Bjvanminnen) [reply dashboard/app/helpers/plc/courses_helper.rb:3]: Pluck was not working after delegating name.

File: dashboard/app/models/pd/workshop.rb (hunk 3 above)
Comment (Bjvanminnen) [reply dashboard/app/models/pd/workshop.rb:484]: `WORKSHOP_COURSE_ONLINE_LEARNING_MAPPING[course]` gives us a string. Just because that string exists, doesn't mean we have a Course of that name.

Comment (aoby) [reply dashboard/app/models/pd/workshop.rb:484]: OIC, I misread. The try is applied to the result of `find_by`. Disregard.

Comment (ashercodeorg) [reply dashboard/app/models/pd/workshop.rb:484]: A priori, both are needed, as the `find_by` can fail to retrieve a record.

**Comment by ashercodeorg** [comment]
It is a little disconcerting to me to see `Plc::Course`, `::Course`, and `Course`. Is this funniness required? How do dumb users like me know which to use, given your comment of environment dependent failures? Is this my simply not knowing enough about Rails (scoping)?
